\documentclass[11pt]{nih}

%%%%%%%%%%%%
% Packages %
%%%%%%%%%%%%

% Spacing and layout
%\usepackage[margin=1in]{geometry}
%\usepackage[utf8]{inputenc}
\usepackage{setspace}

% Math
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{mathtools}
\usepackage{gensymb}

%% Figures
\usepackage{caption}
\usepackage{subcaption}
\usepackage{boxedminipage}

%% Graphics
\usepackage{graphicx}
\usepackage{rotating}
\usepackage{tikz}
\usepackage{color}

% Tables
\usepackage{longtable}
\usepackage{tabularx}

% Hyperlinks
\usepackage{url}
\usepackage[unicode,hidelinks]{hyperref}

% Bibliography
\usepackage[numbers,sort]{natbib}
 
% Misc
\usepackage{enumerate}
\usepackage{titlesec}

% Chemistry
\usepackage[version=3]{mhchem}



%%%%%%%%%%%%
% Commands %
%%%%%%%%%%%%

% Math
\DeclareMathOperator*{\argmin}{arg\,min}
\DeclareMathOperator*{\argmax}{arg\,max}
\newcommand{\Prob}[1]{\Pr\!\left[\,{#1}\,\right]}
\newcommand{\ProbC}[2]{\Prob{{#1}\,|\,{#2}}}

% Dense enumeration
\newenvironment{denseenum}{
\begin{enumerate}
\setlength{\itemsep}{0.1em}
\setlength{\parskip}{0em}
\setlength{\parsep}{0em}
}{\end{enumerate}}


% TODO notes
\newcommand{\todo}[1]{
\addcontentsline{tdo}{todo}{\protect{#1}}
\textcolor{red}{TODO: #1}
}

% TODO list
\makeatletter \newcommand \listoftodos{\section*{\textcolor{red}{Things to do}} \@starttoc{tdo}}
\newcommand\l@todo[2]
{\par\noindent \textcolor{red}{\textbf{#2}: \parbox{10cm}{#1}}\par} \makeatother


%%%%%%%%%%%%%%
% Formatting %
%%%%%%%%%%%%%%

% Spacing between paragraphs, etc.
\setstretch{1.1}
\setlength{\parskip}{0.4em plus0em minus0em}
\setcounter{secnumdepth}{0}
\setlength{\parindent}{0in}
\setlength{\itemsep}{0in}

%\titleformat*{\section}{\Huge\bfseries}
%\titleformat*{\subsection}{\LARGE\bfseries}
%\titleformat*{\subsubsection}{\large\bfseries}





%%%%%%%%%
% Title %
%%%%%%%%%

\title{\textbf{NIH grant proposal}}
\author{Douglas Myers-Turnbull, Robin Betz, Yunsup Jung}
\date{}

\begin{document}

\maketitle

\listoftodos

\tableofcontents


%%%%%%%%%%%%%%%%%
% Specific aims %
%%%%%%%%%%%%%%%%%
\section{Specific aims}

% isoaccepting tRNAs

We propose a large-scale bioinformatics study to identify the effects of synonymous codon usage on protein structure. We intend to address causal relationships rather than statistical associations by developing a mathematically and statistically rigorous framework, which we will use to address a number of hypotheses. \todo{Write a little more here.}
[Robin]

\subsection{Aim 1: Develop a comprehensive, verifiable, and rigorous framework to elucidate relationships between codon usage bias and protein structure.}
We will develop a mathematical and computational framework that will allow the robust detection of relationships between codon usage and variables describing protein structure, even when the overall statistical correlation between the two variables is low.

\subsection{Aim 2: Identify particular effects of synonymous mutations on protein structure.}
We will use the framework established in aim 1 to develop a computational pipeline to detect proteins whose structures are affected by differential codon usage. We hypothesize that many of these structural changes can cause protein misfolding, which may be clinically important.

\subsection{Aim 3: Elucidate general mechanisms that underlie differential codon usage.}
We intend to use the same framework and pipeline described in the previous aims to investigate our hypotheses that codon usage bias is causally related to protein domain organization, secondary structure, knotting, folding environment, and structural complexity. Secondarily, we wish to establish codon usage bias as an important biological mechanism.


%%%%%%%%%%%%%%%%
% Significance %
%%%%%%%%%%%%%%%%
\section{Significance}

\subsection{Existing literature}

\subsubsection{Correlation with expression}
It is known that codon usage bias correlates with expression levels \citep{Goetz2005,Gustafsson2004}. There are strong indications that abundance of isoaccepting tRNA molecules is a causal factor of this differential expression \citep{Klumpp2012,Plotkin2011,Najafabadi2007,Tavare1989}. Proteins with high expression levels  contain greater levels of frequently used codons, and frequently used codons are associated with higher levels of corresponding tRNAs. Thus there is believed to be a positive selection for a small, constrained set of codon--tRNA combinations, and concentrations of codons and tRNAs are related in a positive feedback cycle, where bias in either causes a positive selection for bias in the other. Therefore, statistically significant violations of this general trend are interesting because they indicate the presence of other selection biases. Such selection biases may be at the heart of important mechanisms of protein expression, some which are probably currently unknown.

\subsubsection{Terminology}
Because of strong correlation described above, we call infrequently used codons \emph{slow}, and frequently used codons \emph{fast}, except in cases where we address this correlation directly. Furthermore, we consider sequences containing a large proportion of fast codons to have high \emph{codon usage bias}, and sequences containing either a moderate or low proportion to be \emph{unbiased}, even though the bias of sequences with low proportion still deviate from the statistical mean.

\subsubsection{Importance in translational dynamics}
There is also substantial evidence that codon usage bias is also fundamentally linked to translation dynamics  \citep{Bentele2013,Mitarai2013,Cannarozzi2010,Fredrick2010,Marin2008,Buchan2007}. Factors including mRNA secondary structure \citep{Stoletzki2008,Chamary2005,Baim1985}, mRNA stability \citep{Gu2010}, codon--tRNA affinity \cite{Klumpp2012}, and translational errors \cite{Zhou2009,Drummond2008,Najafabadi2007} have been suggested. In addition, studies have shown that codon usage bias is strongly correlated to ribosomal traffic \citep{Cannarozzi2010a,Buchan2007,Komar1999}.  Particularly, was demonstrated that, for efficient translation, overall codon usage should be skewed in favor of fast codons, and there should be a gradient in which slow codons are prevalent at the beginning of the transcript but rare toward the middle and end. \citet{Mitarai2013} used a simple computational model to show that the introduction of even a single slow codon near the end of the transcript can cause ribosomal traffic jams that drastically decrease translation rate, and presumably also expression. This is believed to be due to ribosomal queueing, a ribosome translating an mRNA transcript interacts physically with the translation process of another ribosome on the same transcript. This prevents both ribosomes from proceeding to subsequent codons.

\subsubsection{Effects on protein structure}
Because slow codons can cause pauses in translation, it has been suggested that synonymous codon usage can influence protein folding, leading to different folded states for transcripts with differing synonymous codon usage \cite{Komar2009,Zhang2009,Buchan2007,Crombie1994}.
Several studies have found that codon usage bias is correlated to protein structure \citep{Saunders2010,Biro2006,Adzhubei1996,Gu2003}, including protein secondary structure \citep{Oresic2003,Gu2003,Thanaraj1996a} and domain organization \citep{Gu2004,Oresic2003}.

%Deana1998 showed regulatory effect
The first known effect of synonymous mutations on protein structure came from \citet{Crombie1994}, who replaced 10 consecutive slow codons with fast codons in the E. coli TRP3 protein. Doing so reduced the native enzymatic activity by 1.5-fold. Still more surprising was the finding by \citet{Sarfaty2007} that a silent polymorphism in the mammalian multi-drug-resistant gene MDR1 altered its folded state and decreased its substrate specificity, without affecting its expression. A recent study by \citet{Zhou2013} found a similar result for a natural mutation in the clinically important circadian rhythm protein FRQ. These studies hint to the existance of more examples of such effects. Furthermore, we argue that additional such cases are likely to be clinically important. 

\subsection{Limitations of existing approaches}

In general, the bioinformatics studies by \citet{Saunders2010,Biro2006,Adzhubei1996,Gu2003} controlled for very few variables and were therefore able to identify only a few clear correlations (most notably with protein secondary structure). However, we hypothesize that these results were negative because the effect of codon usage bias on structure is a weak signal: the effects on most protein structure are minor or nonexistent for most proteins. However, the weakness of the signal does not belie the presence of general mechanisms behind the effects; this is evidence because, as shown by \citet{Zhou2013,Sarfaty2007,Crombie1994}, differential codon usage can dramatically affect the folding of certain proteins. Therefore, we further hypothesize that general mechanisms exist even if few general correlations do, and that controlling for more variables will allow us to elucidate general mechanisms.

\subsection{Further applications}
In addition to clinically significant synonymous mutations and other differential codon usage, our data will have impact on additional applications. It has been recently shown that codon usage bias can be a pivitol factor in de novo protein design []. Although it is know that designed transcripts should contain mostly fast codons, and that there should be a gradient of codon bias along the transcript sequence, potential unwanted effects of synoymous codon usage on a protein product is not generally considered as part of protein design. We therefore note that the discovery of general mechanisms may be important to this application. \todo{At least 1 more additional application.}
[Robin]

%%%%%%%%%%%%%%
% Innovation %
%%%%%%%%%%%%%%
\section{Innovation}

The previous studies by \citet{Saunders2010,Biro2006,Adzhubei1996,Gu2003} examining protein structure in the context of codon usage bias have been constrained to examiniations of statistical correlations. Although they and many other studies [] have suggested that differential codon usage may influence protein structure, such effects have only been demonstrated in vivo by \citet{Zhou2013,Sarfaty2007}.

No studies attempting to investigate causal relationships between codon usage bias and protein structure have been thus far published in peer-reviewed journals, and, to our knowledge, no such investigations have been performed. Therefore, we conclude that our proposal is strictly unique in this endevor.

In addition, the previous studies were partly unsuccessful in establishing even statistical correlations. The investigation by \citet{Saunders2010}, which is the most recent, interrogated differences in codon usage  within helices, strands, and coils by applying both Mantel--Haenszel statistics and a $\chi^2$-test. They found few significant differences between the three secondary structural types, but did find a significant decrease in codon usage bias near the transitions between secondary structural elements. They also investigated the hypothesis that slow codons are frequent around domain boundaries; the results in this case were negative. However, these studies relied on simple statistical techniques that lack the power necessary to find statistical associations for weaker signals.

\todo{Finish}

%%%%%%%%%%%%
% Approach %
%%%%%%%%%%%%
\section{Approach}

%%%%%%%%%%%%%%%%%%%%
% Preliminary data %
%%%%%%%%%%%%%%%%%%%%
%\subsection{Preliminary data}
% Justify our hypotheses

%%%%%%%%%
% Aim 1 %
%%%%%%%%%
\subsection{Aim 1}

%%%%%%%%%%%%%%%%%%%%%%%%%%
% Assessment of accuracy %
%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection{Assessment of accuracy}

%%%%%%%%%
% Aim 2 %
%%%%%%%%%
\subsection{Aim 2}

%%%%%%%%%
% Aim 3 %
%%%%%%%%%
\subsection{Aim 3}

%%%%%%%%%%%%%%%%%
% Folding study %
%%%%%%%%%%%%%%%%%
\subsubsection{Folding study}
As an auxillary study, we will investigate the effects of synonymous mutations in detail for select cases using MD-based folding software. Although de novo folding is still largely unsolved, and de novo folding algorithms are still in their infancy, such an investigation may still reveal insight for some cases that is unavailable through other means. \cite{Zhang:2008p3335}  



%%%%%%%%%%%%%%%%
% Bibliography %
%%%%%%%%%%%%%%%%

\bibliographystyle{plainnat} 
\bibliography{refs}
\appendix


\end{document}


